src/main/java/mekanism/common/content/gear/mekatool/ModuleVeinMiningUnit.java

```diff
@@ -49,30 +51,36 @@ public int getExcavationRange() {
         return excavationRange.get().getRange();
     }
 
-    public static Set<BlockPos> findPositions(BlockState state, BlockPos location, Level world, int maxRange) {
-        Set<BlockPos> found = new LinkedHashSet<>();
-        Set<BlockPos> openSet = new LinkedHashSet<>();
-        openSet.add(location);
-        Block startBlock = state.getBlock();
-        int maxCount = MekanismConfig.gear.disassemblerMiningCount.get() - 1;
-        while (!openSet.isEmpty()) {
-            BlockPos blockPos = openSet.iterator().next();
-            found.add(blockPos);
+    public static Map<BlockPos, BlockState> findPositions(Level world, Map<BlockPos, BlockState> initial, boolean extended, int maxRange) {
+        int maxVein = MekanismConfig.gear.disassemblerMiningCount.get();
+        int maxCount = (int) (initial.size() + maxVein * initial.values().stream().map(bs->bs.getBlock()).distinct().count());
+        Map<BlockPos, BlockState> found = new LinkedHashMap<>();
+        Map<BlockPos, BlockState> openSet = new LinkedHashMap<>();
+        openSet.putAll(initial);
+        for (int dist = 0, next = openSet.size(); !openSet.isEmpty(); next += (found.size() == next && ++dist > 0) ? openSet.size() : 0) {
+            Entry<BlockPos, BlockState> blockEntry = openSet.entrySet().iterator().next();
```

Comment:
If this gets changed to keep track of the iterator:
```java
Iterator<Entry<BlockPos, BlockState>> iterator = openSet.entrySet().iterator();
Entry<BlockPos, BlockState> blockEntry = iterator.next();
```
then instead of calling `openSet.remove(blockPos);` below we can just call `iterator.remove();`